test/distributed/test_store.py

```diff
@@ -311,6 +315,28 @@ def test_multi_worker_with_fixed_world_size(self):
     def test_multi_worker_with_nonfixed_world_size(self):
         self._multi_worker_helper(None)
 
+class StoreBasedBarrierTest(MultiThreadedTestCase):
+    @property
+    def world_size(self):
+        return 1
+
+    def setUp(self):
+        f = tempfile.NamedTemporaryFile()
+        self.__class__.hash_store = dist.HashStore()
+        self.__class__.file_store = dist.FileStore(f.name)
+        self.__class__.tcp_store = dist.TCPStore("localhost", 0, is_master=True)
+        super().setUp()
+        self._spawn_threads()
+
+    def test_hash_store_based_barrier(self):
+        _store_based_barrier(dist.get_rank(), self.hash_store, timedelta(seconds=20))
```

Comment:
We need a test for the timeout case.

Plus we need to lower the minimum way from 10 seconds in c10d to something smaller like 1s to test the looping behavior more aggressively.